File: RepoTransBench/target_projects/Java/C#/l42111996_java-Kcp/tests/original/KcpServerExamplesTest.cs
using System;
using System.Net;
using System.Text;
using Xunit;
using Moq;

// These would be real using statements in an actual port
// using kcp;
// using kcp_example;

namespace KcpOriginalTests
{
    public class KcpServerExamplesTest
    {
        private Mock<IUkcp> mockUkcp;
        private Mock<IMockBuf> mockBuf;

        public KcpServerExamplesTest()
        {
            mockUkcp = new Mock<IUkcp>();
            mockUkcp.Setup(u => u.User).Returns(new MockIUser(new IPEndPoint(IPAddress.Loopback, 12345)));
            mockUkcp.Setup(u => u.GetConv()).Returns(42L);

            mockBuf = new Mock<IMockBuf>();
            mockBuf.Setup(m => m.ReadBytes(It.IsAny<byte[]>())).Callback<byte[]>(b =>
            {
                var src = Encoding.UTF8.GetBytes("hello");
                Array.Copy(src, b, src.Length);
            });
        }

        [Fact]
        public void TestKcp4sharpExampleServerOnConnectedHandleReceiveExceptionAndClose()
        {
            var server = new Kcp4sharpExampleServer();

            server.OnConnected(mockUkcp.Object);

            var ex = new Exception("error");

            Exception receiveEx = Record.Exception(() => server.HandleReceive(mockBuf.Object, mockUkcp.Object));
            Assert.Null(receiveEx);

            Exception handleEx = Record.Exception(() => server.HandleException(ex, mockUkcp.Object));
            Assert.Null(handleEx);

            Exception closeEx = Record.Exception(() => server.HandleClose(mockUkcp.Object));
            Assert.Null(closeEx);

            mockUkcp.Verify(u => u.Write(It.IsAny<object>()), Times.AtLeastOnce());
        }

        [Fact]
        public void TestKcpDisconnectExampleServerOnConnectedHandleReceiveExceptionAndClose()
        {
            var server = new KcpDisconnectExampleServer();

            server.OnConnected(mockUkcp.Object);

            var ex = new Exception("error");

            Exception receiveEx = Record.Exception(() => server.HandleReceive(mockBuf.Object, mockUkcp.Object));
            Assert.Null(receiveEx);

            Exception handleEx = Record.Exception(() => server.HandleException(ex, mockUkcp.Object));
            Assert.Null(handleEx);

            Exception closeEx = Record.Exception(() => server.HandleClose(mockUkcp.Object));
            Assert.Null(closeEx);

            mockUkcp.Verify(u => u.Write(It.IsAny<object>()), Times.AtLeastOnce());

        }

        [Fact]
        public void TestKcpMultiplePingPongExampleServerLifecycle()
        {
            var server = new KcpMultiplePingPongExampleServer();

            server.OnConnected(mockUkcp.Object);

            var ex = new Exception("error2");

            Exception receiveEx = Record.Exception(() => server.HandleReceive(mockBuf.Object, mockUkcp.Object));
            Assert.Null(receiveEx);

            Exception handleEx = Record.Exception(() => server.HandleException(ex, mockUkcp.Object));
            Assert.Null(handleEx);

            Exception closeEx = Record.Exception(() => server.HandleClose(mockUkcp.Object));
            Assert.Null(closeEx);

            mockUkcp.Verify(u => u.Write(It.IsAny<object>()), Times.AtLeastOnce());

        }

        [Fact]
        public void TestKcpReconnectExampleServerAllPaths()
        {
            var server = new KcpReconnectExampleServer();

            server.OnConnected(mockUkcp.Object);

            server.HandleReceive(mockBuf.Object, mockUkcp.Object);
            // Simulate time passing to hit other path
            server.Start = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - 2000;
            server.HandleReceive(mockBuf.Object, mockUkcp.Object);

            var ex = new Exception("error3");
            Exception handleEx = Record.Exception(() => server.HandleException(ex, mockUkcp.Object));
            Assert.Null(handleEx);

            Exception closeEx = Record.Exception(() => server.HandleClose(mockUkcp.Object));
            Assert.Null(closeEx);

            mockUkcp.Verify(u => u.Write(It.IsAny<object>()), Times.AtLeastOnce());
        }

        [Fact]
        public void TestSpeedExampleServerFlow()
        {
            var server = new SpeedExampleServer();

            server.OnConnected(mockUkcp.Object);

            Exception receiveEx = Record.Exception(() => server.HandleReceive(mockBuf.Object, mockUkcp.Object));
            Assert.Null(receiveEx);

            server.Start = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - 1200;

            Exception receiveEx2 = Record.Exception(() => server.HandleReceive(mockBuf.Object, mockUkcp.Object));
            Assert.Null(receiveEx2);

            var ex = new Exception("error4");
            Exception handleEx = Record.Exception(() => server.HandleException(ex, mockUkcp.Object));
            Assert.Null(handleEx);

            Exception closeEx = Record.Exception(() => server.HandleClose(mockUkcp.Object));
            Assert.Null(closeEx);
        }
    }

    // Mocks and interface stubs for porting, real implementations must be provided per actual library
    public interface IUkcp
    {
        IMockUser User { get; }
        long GetConv();
        void Write(object obj);
    }
    public class MockIUser : IMockUser
    {
        private IPEndPoint _remoteAddress;
        public MockIUser(IPEndPoint remoteAddress) { _remoteAddress = remoteAddress; }
        public IPEndPoint GetRemoteAddress() => _remoteAddress;
    }
    public interface IMockUser
    {
        IPEndPoint GetRemoteAddress();
    }

    public interface IMockBuf
    {
        void ReadBytes(byte[] buf);
    }

    // Stubs for the test servers, with property Start for time simulation
    public class Kcp4sharpExampleServer
    {
        public void OnConnected(IUkcp ukcp) { ukcp.Write("connected"); }
        public void HandleReceive(IMockBuf buf, IUkcp ukcp) { ukcp.Write("received"); }
        public void HandleException(Exception ex, IUkcp ukcp) { ukcp.Write("exception"); }
        public void HandleClose(IUkcp ukcp) { ukcp.Write("closed"); }
    }
    public class KcpDisconnectExampleServer : Kcp4sharpExampleServer { }
    public class KcpMultiplePingPongExampleServer : Kcp4sharpExampleServer { }
    public class KcpReconnectExampleServer : Kcp4sharpExampleServer
    {
        public long Start { get; set; }
        public new void HandleReceive(IMockBuf buf, IUkcp ukcp)
        {
            // Simulates bifurcated behavior based on time (like original)
            if ((DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - Start) > 1000)
                ukcp.Write("another path");
            else
                ukcp.Write("normal");
        }
    }
    public class SpeedExampleServer : Kcp4sharpExampleServer
    {
        public long Start { get; set; }
        public new void HandleReceive(IMockBuf buf, IUkcp ukcp)
        {
            // Simulates two behaviors (based on time)
            if ((DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - Start) > 1000)
                ukcp.Write("time passed");
            else
                ukcp.Write("first");
        }
    }
}

File: RepoTransBench/target_projects/Java/Go/l42111996_java-Kcp/tests/original/kcp_server_examples_test.go
package original

import (
	"errors"
	"net"
	"testing"
	"time"
)

// Create mock ukcp and buffer types for the test

type MockUkcp struct {
	writeInvoked int
	userData     *MockUserData
}

type MockUserData struct {
	remoteAddr net.Addr
}

func (u *MockUkcp) user() *MockUserData {
	return u.userData
}

func (u *MockUkcp) getConv() uint64 {
	return 42
}

func (u *MockUkcp) write(_ interface{}) {
	u.writeInvoked++
}

type MockBuf struct {
	data []byte
}

func (b *MockBuf) WriteBytes(d []byte) {
	b.data = append(b.data, d...)
}

func newMockBuf() *MockBuf {
	b := &MockBuf{}
	b.WriteBytes([]byte("hello"))
	return b
}

// Mocks for the various servers

type Kcp4sharpExampleServer struct{}
type KcpDisconnectExampleServer struct{}
type KcpMultiplePingPongExampleServer struct{}
type KcpReconnectExampleServer struct {
	start int64
}
type SpeedExampleServer struct {
	start int64
}

func (Kcp4sharpExampleServer) onConnected(ukcp *MockUkcp)                           { ukcp.write("ok") }
func (Kcp4sharpExampleServer) handleReceive(buf *MockBuf, ukcp *MockUkcp) error     { ukcp.write(buf.data); return nil }
func (Kcp4sharpExampleServer) handleException(_ error, ukcp *MockUkcp) error        { ukcp.write("exception"); return nil }
func (Kcp4sharpExampleServer) handleClose(ukcp *MockUkcp) error                     { ukcp.write("closed"); return nil }
func (KcpDisconnectExampleServer) onConnected(ukcp *MockUkcp)                       { ukcp.write("ok") }
func (KcpDisconnectExampleServer) handleReceive(buf *MockBuf, ukcp *MockUkcp) error { ukcp.write(buf.data); return nil }
func (KcpDisconnectExampleServer) handleException(_ error, ukcp *MockUkcp) error    { ukcp.write("exception"); return nil }
func (KcpDisconnectExampleServer) handleClose(ukcp *MockUkcp) error                 { ukcp.write("closed"); return nil }
func (KcpMultiplePingPongExampleServer) onConnected(ukcp *MockUkcp)                 { ukcp.write("ok") }
func (KcpMultiplePingPongExampleServer) handleReceive(buf *MockBuf, ukcp *MockUkcp) error {
	ukcp.write(buf.data)
	return nil
}
func (KcpMultiplePingPongExampleServer) handleException(_ error, ukcp *MockUkcp) error { ukcp.write("exception"); return nil }
func (KcpMultiplePingPongExampleServer) handleClose(ukcp *MockUkcp) error              { ukcp.write("closed"); return nil }

func (s *KcpReconnectExampleServer) onConnected(ukcp *MockUkcp) { ukcp.write("ok"); s.start = time.Now().UnixMilli() }
func (s *KcpReconnectExampleServer) handleReceive(buf *MockBuf, ukcp *MockUkcp) error {
	ukcp.write(buf.data)
	return nil
}
func (s *KcpReconnectExampleServer) handleException(_ error, ukcp *MockUkcp) error { ukcp.write("exception"); return nil }
func (s *KcpReconnectExampleServer) handleClose(ukcp *MockUkcp) error              { ukcp.write("closed"); return nil }

func (s *SpeedExampleServer) onConnected(ukcp *MockUkcp) { ukcp.write("ok"); s.start = time.Now().UnixMilli() }
func (s *SpeedExampleServer) handleReceive(buf *MockBuf, ukcp *MockUkcp) error {
	ukcp.write(buf.data)
	return nil
}
func (s *SpeedExampleServer) handleException(_ error, ukcp *MockUkcp) error { ukcp.write("exception"); return nil }
func (s *SpeedExampleServer) handleClose(ukcp *MockUkcp) error              { ukcp.write("closed"); return nil }

func TestKcp4sharpExampleServerOnConnectedHandleReceiveExceptionAndClose(t *testing.T) {
	ukcp := &MockUkcp{userData: &MockUserData{remoteAddr: &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 12345}}}
	buf := newMockBuf()
	server := Kcp4sharpExampleServer{}
	server.onConnected(ukcp)
	if err := server.handleReceive(buf, ukcp); err != nil {
		t.Errorf("handleReceive failed: %v", err)
	}
	err := errors.New("error")
	if err2 := server.handleException(err, ukcp); err2 != nil {
		t.Errorf("handleException failed: %v", err2)
	}
	if err2 := server.handleClose(ukcp); err2 != nil {
		t.Errorf("handleClose failed: %v", err2)
	}
	if ukcp.writeInvoked < 1 {
		t.Errorf("expected ukcp write to be called")
	}
}

func TestKcpDisconnectExampleServerOnConnectedHandleReceiveExceptionAndClose(t *testing.T) {
	ukcp := &MockUkcp{userData: &MockUserData{remoteAddr: &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 12345}}}
	buf := newMockBuf()
	server := KcpDisconnectExampleServer{}
	server.onConnected(ukcp)
	if err := server.handleReceive(buf, ukcp); err != nil {
		t.Errorf("handleReceive failed: %v", err)
	}
	err := errors.New("error")
	if err2 := server.handleException(err, ukcp); err2 != nil {
		t.Errorf("handleException failed: %v", err2)
	}
	if err2 := server.handleClose(ukcp); err2 != nil {
		t.Errorf("handleClose failed: %v", err2)
	}
	if ukcp.writeInvoked < 1 {
		t.Errorf("expected ukcp write to be called")
	}
}

func TestKcpMultiplePingPongExampleServerLifecycle(t *testing.T) {
	ukcp := &MockUkcp{userData: &MockUserData{remoteAddr: &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 12345}}}
	buf := newMockBuf()
	server := KcpMultiplePingPongExampleServer{}
	server.onConnected(ukcp)
	if err := server.handleReceive(buf, ukcp); err != nil {
		t.Errorf("handleReceive failed: %v", err)
	}
	err := errors.New("error2")
	if err2 := server.handleException(err, ukcp); err2 != nil {
		t.Errorf("handleException failed: %v", err2)
	}
	if err2 := server.handleClose(ukcp); err2 != nil {
		t.Errorf("handleClose failed: %v", err2)
	}
	if ukcp.writeInvoked < 1 {
		t.Errorf("expected ukcp write to be called")
	}
}

func TestKcpReconnectExampleServerAllPaths(t *testing.T) {
	ukcp := &MockUkcp{userData: &MockUserData{remoteAddr: &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 12345}}}
	buf := newMockBuf()
	server := &KcpReconnectExampleServer{}
	server.onConnected(ukcp)
	_ = server.handleReceive(buf, ukcp)
	// Simulate time passing to hit the other path
	server.start = time.Now().UnixMilli() - 2000
	_ = server.handleReceive(buf, ukcp)
	ex := errors.New("error3")
	if err := server.handleException(ex, ukcp); err != nil {
		t.Errorf("handleException failed: %v", err)
	}
	if err := server.handleClose(ukcp); err != nil {
		t.Errorf("handleClose failed: %v", err)
	}
	if ukcp.writeInvoked < 1 {
		t.Errorf("expected ukcp write to be called")
	}
}

func TestSpeedExampleServerFlow(t *testing.T) {
	ukcp := &MockUkcp{userData: &MockUserData{remoteAddr: &net.TCPAddr{IP: net.ParseIP("127.0.0.1"), Port: 12345}}}
	buf := newMockBuf()
	server := &SpeedExampleServer{}
	server.onConnected(ukcp)
	if err := server.handleReceive(buf, ukcp); err != nil {
		t.Errorf("handleReceive failed: %v", err)
	}
	server.start = time.Now().UnixMilli() - 1200
	if err := server.handleReceive(buf, ukcp); err != nil {
		t.Errorf("handleReceive failed (second): %v", err)
	}
	ex := errors.New("error4")
	if err := server.handleException(ex, ukcp); err != nil {
		t.Errorf("handleException failed: %v", err)
	}
	if err := server.handleClose(ukcp); err != nil {
		t.Errorf("handleClose failed: %v", err)
	}
}